### PR TITLE
Firewall rules deployment strategy with HDFS DataNode as example

### PR DESCRIPTION
**Pull request by Nuttymoon**
#### Which issue(s) this PR fixes:

Partially fixes `#422`

#### Additional comments:

**This PR is a follow up on PR #452 and #481**

**Work in progress:** the firewall rules are only implemented for HDFS DataNode. The goal of this PR for now is to get feedback on the implementation.

The implementation proposed here inherits the concepts of [firewalld](https://firewalld.org/): [zones](https://firewalld.org/documentation/zone/) and [services](https://firewalld.org/documentation/service/). A service corresponds to a list of ports (+protocols) and a zone corresponds to a list of sources (~= IPs) and the rules that apply to them when accessing services on the host.

##### Possible negative impacts

Deploying erroneous firewall rules on hosts can have disastrous consequences, notably forbidding node operators to connect via SSH. I suggest that we implement a check that determines if Ansible will still be able to connect after the firewall rules deployment and if not abort it (see #480).

##### `firewall` role

A new `firewall` role is added to TDP. This role will always have to be run prior to services deployment. The role has been written to support multiple firewall services as a backend. For now only `firewalld` is supported. One can change which firewall service is used with `firewall_service` and the routing logic is managed in `firewall.generic`.

```
roles/firewall
├── firewalld
│   ├── handlers
│   ├── tasks
│   └── templates
└── generic
    └── tasks
```

##### Cluster level firewall

`firewall_tdp_managed` controls if TDP should generate firewall rules on the cluster nodes:
- If `yes`, TDP ensures that all rules defined in `firewall_rules` are created on the right hosts and that `firewalld` is started.
- If `no`, TDP does not interfere with `firewalld`.

https://github.com/TOSIT-IO/tdp-collection/blob/14809dd1a1c8c91778bf5ebe99b60a7daa0649fb/tdp_vars_defaults/firewall/firewall.yml#L5-L10

##### Firewall zones

By default, TDP will create 2 zones:

- `public`: This zone corresponds to all the hosts outside of the TDP cluster that need access to its client endpoints (e.g. HDFS Web UI, Hive JDBC/ODBC).
- `internal`: This zone corresponds to all the TDP cluster hosts. Internal ports need to be opened inside the cluster.

https://github.com/TOSIT-IO/tdp-collection/blob/14809dd1a1c8c91778bf5ebe99b60a7daa0649fb/tdp_vars_defaults/firewall/firewall.yml#L16-L26

Each zone has 4 properties:

- `name`: A name for the zone
- `short`: A short description of the rule that will be used in the `firewalld` zone definition
- `groups`: The Ansible groups that are part of this zone. TDP will take care of mapping these groups to [firewalld ipsets](https://firewalld.org/documentation/ipset/).
- `addresses`: The other addresses that are part of this zone. Can be in the format `address` or `address/mask`.

##### Firewall rules

All open port rules are listed in the `firewall_rules` variable, with one list for each component. E.g. for `hdfs_dn` (HDFS DataNode):

https://github.com/TOSIT-IO/tdp-collection/blob/14809dd1a1c8c91778bf5ebe99b60a7daa0649fb/tdp_vars_defaults/firewall/firewall.yml#L40-L48

Each rule has 5 properties:

- `short`: A short description of the rule that will be used in the `firewalld` service definition
- `port`: The port to open and secure
- `protocol`: The protocol authorized on this port
- `zones`: The zones from which the port will be accessible.
- `state`: `present` or `absent`. If `absent`, TDP will make sure that the port is removed from all rules.

##### Trying this out

If you want to try this on a TDP cluster, run `playbooks/meta/firewall.yml`. This will deploy firewall rules on `hdfs_dn` hosts and start `firewalld`. Note that other services installed on the same hosts (e.g. YARN RM) will stop functioning as no firewall rules will exist for them.

##### Low level implementation

TDP uses `firewalld` ipsets, services, and zones to set firewall rules (all generated files are prefixed with `firewalld_files_prefix`):

- **Firewalld ipsets:** each Ansible group that is used in `firewall_zones` will be mapped to an ipset in `/etc/firewalld/ipsets`. E.g. `/etc/firewalld/ipsets/tdp_tdp_cluster.xml`:

  ```xml
  <?xml version="1.0" encoding="utf-8"?>
  <ipset type="hash:net">
    <short>tdp_cluster Ansible group</short>
    <entry>192.168.56.13</entry>
    ...
    <entry>192.168.56.16</entry>
  </ipset>
  ```

- **Firewalld services:** each TDP component has 1 service (in `/etc/firewalld/services`) for each of its `firewalld_rules`. A service contains the port number and protocol. E.g. `/etc/firewalld/services/tdp_hdfs_dn_data_port.xml`:

  ```xml
  <?xml version="1.0" encoding="utf-8"?>
  <service>
    <short>HDFS DataNode data transfer port</short>
    <port protocol="tcp" port="9866" />
  </service>
  ```

- **Firewalld zones:** Each zone in `firewall_zones` has 1 zone in `/etc/firewalld/zones` that defines the sources allowed to the zone's services. E.g. `/etc/firewalld/zones/tdp_internal.xml`:

  ```xml
  <?xml version='1.0' encoding='UTF-8'?>
  <zone>
    <short>TDP cluster internal zone</short>
    <source ipset="tdp_tdp_cluster"/>
    <service name="tdp_hdfs_dn_data_port"/>
    <service name="tdp_hdfs_dn_ipc_port"/>
    ...
  </zone>
  ```

#### License agreement

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: tdp_vars_defaults/firewall/firewall.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+firewalld_rules:
+  # Rules created on all hosts
+  tdp_cluster:
+    - name: ssh
+      short: SSH port
+      port: 22
+      protocol: tcp
+      zones:
+        - public
+      state: present
```
Comment (leopaul36) [inline tdp_vars_defaults/firewall/firewall.yml:38]: As discussed before, I am not really comfortable with TDP managing rules outside of the TDP scope.

Also in this particular case, I think `ssh` is in the public zone by default in firewalld installations:

```
firewall-cmd --list-services --zone=public 
dhcpv6-client ssh
```

Comment (Nuttymoon) [reply tdp_vars_defaults/firewall/firewall.yml:38]: I wanted to cover the edge case where `firewalld` is not pre-configured on the host (I think this is the case for LXC instances). But I am fine with removing this rule.

Comment (leopaul36) [reply tdp_vars_defaults/firewall/firewall.yml:38]: I tested with LXD images `ubuntu:` and `images:rockylinux/8`, both had ssh by default in the public zone

Comment (Nuttymoon) [reply tdp_vars_defaults/firewall/firewall.yml]: I removed this section

File: tdp_vars_defaults/firewall/firewall.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2022 TOSIT.IO
+# SPDX-License-Identifier: Apache-2.0
+
+---
+# Should TDP manage firewall rules on cluster nodes
+firewall_tdp_managed: yes
```
Comment (rpignolet) [inline tdp_vars_defaults/firewall/firewall.yml:6]: As discuss, change to `no` by default.

**Comment by rpignolet** [comment]
`meta/all.yml` is missing

**Comment by Nuttymoon** [comment]
> `meta/all.yml` is missing

Done

**Comment by Nuttymoon** [comment]
Force pushed after rebasing `422-all-firewall-rules` on `master`

**Comment by Nuttymoon** [comment]
@leopaul36 @gboutry @mehdibn are you ok to merge this?

As discussed Friday, this PR does not go to the master branch so we can still discuss where to put the firewall rules later on (e.g. move them to a dedicated collection or to `tdp-extras`).

**Comment by leopaul36** [comment]
> @leopaul36 @gboutry @mehdibn are you ok to merge this?
> 
> As discussed Friday, this PR does not go to the master branch so we can still discuss where to put the firewall rules later on (e.g. move them to a dedicated collection or to `tdp-extras`).

Ok for me

File: roles/firewall/generic/tasks/start.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Copyright 2022 TOSIT.IO
+# SPDX-License-Identifier: Apache-2.0
+
+---
+- name: "Start {{ firewall_service }}"
+  service:
+    name: "{{ firewall_service }}"
+    state: started
```
Comment (rpignolet) [inline roles/firewall/generic/tasks/start.yml:8]: `enabled` is missing no ?

Comment (Nuttymoon) [reply roles/firewall/generic/tasks/start.yml:8]: I enable the service in the `config` operation
https://github.com/TOSIT-IO/tdp-collection/blob/8aed16ea280086f4c8bbc8f40231aeacce46d05b/roles/firewall/firewalld/tasks/config.yml#L54

But I can put it in the `start/restart/stop` to uniformize with other components indeed.

**Review by rpignolet — APPROVED** [review]

**Pull request merged**